### ISSUE TITLE
Success flag and FluxMaps/FluxPoints

### ISSUE THREAD

**Issue by registerrier** [gammapy/gammapy#3493]
The success flag is no longer present in the `FluxPoints`. It can be important to track errors.

Should we add a boolean `success` quantity in `FluxMaps`? Or should we put e.g. `nan` values where optimization/confidence has failed?

Opinions @adonath , @QRemy , @bkhelifi @AtreyeeS ?

**Comment by AtreyeeS** [comment]
I vote for `success` flags rather than `nan`

**Comment by bkhelifi** [comment]
I have a preference for a NaN, as one could not check the content of all bins in a FluxMap. In the plot function, this will screw up everything, which is great as it will tell to the user that something went wrong. But maybe one could add an argument to remove the non-converged bins...

One could add a boolean flag additionally, such that the user can understand the reason of this NaN. One might have NaN coming from other reasons (or maybe not).

PS: it is sad that the documentation of `FluxEstimator` is not visible, as the code to compute flux in `FluxPointsEstimator` is actually in `FluxEstimator`
PS2: the same for the documentation of `FluxMaps` that is not visible...
PS2: this is an issue with the documentation of the function sqrt_ts of `FluxPoints`.

**Comment by QRemy** [comment]
I would prefer to keep the bad values rather than using nan as they often contain information on what went wrong by themself (stuck at best fit-value or min/max bounds, errors missing or way too small/large...).

 If we provide the convergence status as a boolean array the users can easily remove them in one line  if they want to: `FluxMap[convergence_failure]=np.nan`. 
 
Moreover we could have one mask for covariance_failure and one array for convergence_failure with different values depending on the success message as convergence_failure = 0:success, 1:estimated distance to minimum (EDM) too small, 2 failure. So we can decide to keep or not the EDM too small case that often correspond to a success.

**Comment by registerrier** [comment]
The good thing with having `nan` is that `Map.plot` will directly plot a blank entry. So that if you do `flux_maps.dnde.plot()` you won't see the pixels without convergence.

This behavior can be achieved with the approach suggested by @QRemy if we modify the various flux properties of `FluxMaps` e.g.:
```
 def _filter_convergence_failure(self, some_map):
     """Put NaN where pixels did not converge."""
     some_map[~self.convergence_success.data] = np.nan
     return some_map

@property
def dnde(self):
"""Return differential flux (dnde) SED values."""
    return self._filter_convergence_failure(self.norm * self.dnde_ref)
```

**Comment by adonath** [comment]
I think that's an interesting suggestion to consider @registerrier and @QRemy! Just a few more thoughts from my side:
- Typically we already need to set `np.nan` on a lower level, i.e. the `FluxEstimator`level. There are cases where the fit just completely fails with an error and this case one just has to set `NaN`, because there is no other information to put. 
- For the Gammapy API we should keep a more complex data structure and not a simple flag. As proposed by @QRemy this could either be an integer status flag (this also proposed in GADF as `fit_status` in the [TS cube format](https://gamma-astro-data-formats.readthedocs.io/en/latest/spectra/binned_likelihoods/index.html#fitdata-table)). Or we could even store the complete `FitResult` object, which is useful for interactive inspection. On serialisation this can be reduced to an integer number of boolean value. 
- The advantage of a more flexible data structure is that things like `niter` or marginalised parameters can be stored as well.

**Comment by registerrier** [comment]
> * For the Gammapy API we should keep a more complex data structure and not a simple flag. As proposed by @QRemy this could either be an integer status flag (this also proposed in GADF as `fit_status` in the [TS cube format](https://gamma-astro-data-formats.readthedocs.io/en/latest/spectra/binned_likelihoods/index.html#fitdata-table)). Or we could even store the complete `FitResult` object, which is useful for interactive inspection. On serialisation this can be reduced to an integer number or boolean value.

Is there an easy scheme to store something like the `FitResult`? I suppose we can define a `Map` with an object `dtype` so that we could store `FitResult` objects. Have anybody tried that? Can it work?

**Comment by adonath** [comment]
@registerrier I haven't tried, but yes it should work at least for storage of data. I think the only alternative approach is to store arrays in `FitResult` attributes, such as `FitResult.success`, but this is less intuitive and one would need a scheme to access the individual values.

**Comment by adonath** [comment]
Here is a minimal example:
```python3
from gammapy.modeling.fit import FitResult
from gammapy.maps import RegionNDMap, MapAxis
import numpy as np
from astropy import units as u

axis = MapAxis.from_energy_bounds("1 TeV", "10 TeV", nbin=5)

results = []

for idx in range(axis.nbin):
    result = FitResult(
        parameters=[1, 2, 3],
        backend="imimuit",
        method="minos",
        success=True,
        message="Great fit!"
    )
    results.append(result)

fit_results = RegionNDMap.create(
    region="galactic;circle(0, 0, 1)", axes=[axis]
)

fit_results.data = np.array(results)

result = fit_results.slice_by_idx({"energy": 3}).data[0, 0]
print(result)
```
It's not super convenient, but technically it works...

I'm not sure the `Map` object offers a big advantage in this case, but to make it uniform with the other quantities it could be a reasonable choice.

**Comment by bkhelifi** [comment]
I have a naive question. With this option, what is the relative increase of memory size? For the GPS, does this option might be problematic with the large maps?

**Comment by adonath** [comment]
@bkhelifi I think storing the fit results can always be optional. It's mostly useful for debugging, however it could lead us into a direction where we store the individual quantities such as `success`, `niter` in `Map` objects as well, instead of storing the `FitResult` object. I remember looking e.g. at the `niter` map returned by the `TSMapEstimator` in rare cases and use `Map.plot()` to visualise it. This would not work when we store a `FitResult` object. 

So a possible solution could be to store a subset of the information in the `FitResult` object directly on the `FluxMaps` object.

This could include:
- `fit_status` (an integer with the status)
- `success` (a single boolean flag, derived from `fit_status` using some criterion)
- `niter` (number of function evaluations)
- anything else?

**Comment by bkhelifi** [comment]
If I remember correctly, from iminuit, minos/migrad can converge but not hess. So maybe a fit_status (int) is the best... 
For scipy, the return function contains a bool and a message. Do we read or interpret the message? If not, a bool or a 0/1 is sufficient.
For sherpa (https://cxc.cfa.harvard.edu/sherpa/ahelp/get_fit_results.html), there is a bool and a message. So it depends also if we read this message

**Comment by adonath** [comment]
The initial issue is addressed by #3544. For further discussion on a more general `fit_status` we can use directly use #3518, which has already a prototype implementation.
